File: Python/contains_duplicate_2.py
# https://leetcode.com/problems/contains-duplicate-ii/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def containsNearbyDuplicate(self, nums: list[int], k: int) -> bool:
        map_: dict[int, int] = {}
        for i, num in enumerate(nums):
            if num not in map_:
                map_[num] = i
                continue
            if k >= i - map_[num]:
                return True
            map_[num] = i
        return False

def examine()->None:
    Examiner(Solution(), 'containsNearbyDuplicate').run((
        Testcase(([1,2,3,1], 3), True),
        Testcase(([1,0,1,1], 1), True),
        Testcase(([1,2,3,1,2,3], 2), False),
    ))

File: TypeScript/solutions/contains-duplicate-2.ts
// https://leetcode.com/problems/contains-duplicate-ii/
import { Examiner, Testcase } from '../utils';


function containsNearbyDuplicate(nums: number[], k: number): boolean {
    const map = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        const num = nums[i]!;
        if (!map.has(num)) { map.set(num, i); continue; }
        if (k >= i - map.get(num)!) return true;
        map.set(num, i);
    }
    return false;
}

export function examine(): void {
    new Examiner(containsNearbyDuplicate).run([
        new Testcase([[1, 2, 3, 1], 3], true),
        new Testcase([[1, 0, 1, 1], 1], true),
        new Testcase([[1, 2, 3, 1, 2, 3], 2], false),
    ]);
}
